### PR TITLE
Refactor flutter.js to do dart2wasm bootstrapping and CanvasKit/Skwasm preloading.

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
This PR makes some major revisions to our flutter.js bootstrapper.
* Modularize flutter.js into multiple files to make it a little simpler to manage from source code. They are still bundled into a single .js file by esbuild.
* Added a `types.d.ts` file which contains declarations of the types of some of the objects used in the flutter.js API
* Deprecated the old `FlutterLoader.loadEntrypoint` API and added a new function simply called `FlutterLoader.load`, which has a few more capabilities:
  -  A build tool can inject a build config, that may describe multiple builds that `FlutterLoader.load` can attempt to use. It will use the first one that is compatible with the browser environment and the user's configuration.
  - It can also load wasm flutter apps.
  - It also pre-loads and instantiates CanvasKit (and Skwasm) as necessary depending on the build configuration.
  - `FlutterLoader.load` also immediately takes a flutter configuration object. If an `onEntrypointLoaded` callback is not provided by the user, it just does the expected thing and initializes the engine and immediately starts the app, passing the configuration along as needed.
 * `flutter.js` has the engine hash built into it now, which allows it to ascertain the correct CDN URLs for both CanvasKit and Skwasm.

### CODE REVIEW

File: lib/web_ui/flutter_js/BUILD.gn [hunk 1]
```diff
@@ -18,6 +20,17 @@ copy("flutter_js_sources") {
...
 }
 
+generated_file("flutter_js_engine_version") {
+  outputs = [ "$root_out_dir/flutter_web_sdk/flutter_js/src/revision.js" ]
+  contents = [
+    "// Copyright 2013 The Flutter Authors. All rights reserved.",
+    "// Use of this source code is governed by a BSD-style license that can be",
+    "// found in the LICENSE file.",
+    "",
+    "export const engineRevision = \"$engine_version\";",
```
Comment (yjbanov) [inline lib/web_ui/flutter_js/BUILD.gn:30]: This scares me for a couple of reasons:

* The output of web build rules depends on every byte in the engine codebase. For example, changing a `README` file will result in altering the git revision and invalidating the build.
* Git revision does not fully describe the engine revision. You can have local source changes while the git revision is the same.

This probably invalidates one of the core principles of a build system, but I'm failing to name one, maybe hermeticity.

File: lib/web_ui/flutter_js/src/browser_environment.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    // Frequently, when a browser launches an API that other browsers already
+    // support, there are subtle incompatibilities that may cause apps to crash if,
+    // we blindly adopt the new implementation. This check prevents us from picking
+    // up potentially incompatible implementations of ImagdeDecoder API. Instead,
+    // when a new browser engine launches the API, we'll evaluate it and enable it
+    // explicitly.
+    if (!isBlink()) {
+        return false;
+    }
+    return true;
```
Comment (yjbanov) [inline lib/web_ui/flutter_js/src/browser_environment.js:36]: `return isBlink()`?

File: lib/web_ui/flutter_js/src/browser_environment.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
Comment (yjbanov) [inline lib/web_ui/flutter_js/src/browser_environment.js:1]: Other files use two-space indent, but this one uses four spaces.

File: lib/web_ui/flutter_js/BUILD.gn (hunk 1 above)
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/BUILD.gn:30]: Yeah this is an interesting one, I get your concerns. Could we maybe have a GVC and chat about the subtleties here at some point if you have a moment?

Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/BUILD.gn:30]: We had some discussion in a GVC. I think I mostly agree that it would be better to leave the engine revision out of `flutter.js` and instead have the tool inject the engine hash as part of the build config.

File: lib/web_ui/flutter_js/src/flutter.js
Comment (ditman) [inline lib/web_ui/flutter_js/src/flutter.js:1]: Beautiful!

File: lib/web_ui/flutter_js/src/instantiate_wasm.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// This is a little helper function that helps us start the fetch and compilation
+// of an emscripten wasm module in parallel with the fetch of its script.
+export const createWasmInstantiator = (url) => {
+  const modulePromise = WebAssembly.compileStreaming(fetch(url));
+  return (imports, successCallback) => {
```
Comment (ditman) [inline lib/web_ui/flutter_js/src/instantiate_wasm.js:9]: This `successCallback` here is a little bit weird in an async function, shouldn't you be able to:

```js
const modulePromise = WebAssembly.compileStreaming(fetch(url));
return (imports) => new Promise((resolve, reject) => {
  modulePromise.then((module) => {
    WebAssembly.instantiate(module, imports).then((instance) => {
      resolve({
        instance: instance,
        module: module,
      });
    });
  });
});
```

And from the outside you can:

```js
const { module, instance } = await wasmInstantiator(url)(imports); // Very Functional!
```

(I'm sure this can be simplified even further but my 🧠 is 🍳)

Comment (ditman) [reply lib/web_ui/flutter_js/src/instantiate_wasm.js:9]: (I wonder if the promise resolver can be an async function so it can be implemented with `await` rather than `.then`)

File: lib/web_ui/flutter_js/src/loader.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  /**
+   * @deprecated Use `load` instead.
+   * Initializes the Flutter web app.
+   * @param {*} options
+   * @returns {Promise?} a (Deprecated) Promise that will eventually resolve
+   *                     with an EngineInitializer, or will be rejected with
+   *                     any error caused by the loader. Or Null, if the user
+   *                     supplies an `onEntrypointLoaded` Function as an option.
+   */
+  async loadEntrypoint(options) {
```
Comment (ditman) [inline lib/web_ui/flutter_js/src/loader.js:28]: Is `load` a superset of `loadEntrypoint`? Can `loadEntrypoint` be re-implemented as a config+call to `load`?

File: lib/web_ui/flutter_js/src/skwasm_loader.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+            // When hosted via a CDN or some other url that is not the same
+            // origin as the main script of the page, we will fail to create
+            // a web worker with the .worker.js script. This workaround will
+            // make sure that the worker JS can be loaded regardless of where
+            // it is hosted.
+            const url = scriptDirectory + fileName;
+            if (url.endsWith('.worker.js')) {
+              return URL.createObjectURL(new Blob(
+                [`importScripts('${url}');`],
+                { 'type': 'application/javascript' }));
```
Comment (ditman) [inline lib/web_ui/flutter_js/src/skwasm_loader.js:32]: eyes-on-fire.gif

File: lib/web_ui/flutter_js/src/skwasm_loader.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+          }
+        });
+        resolve(skwasmInstance);
+      } catch (e) {
+        reject(e);
+      }
+    });
+    script.addEventListener('error', (e) => {
+      reject(e);
+    });
```
Comment (ditman) [inline lib/web_ui/flutter_js/src/skwasm_loader.js:44]: Can be also written as:

```suggestion
    script.addEventListener('error', reject);
```

File: lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart [hunk 8]
```diff
@@ -43,6 +43,9 @@ external set windowFlutterCanvasKit(CanvasKit? value);
 @JS('window.flutterCanvasKit')
 external CanvasKit? get windowFlutterCanvasKit;
 
+@JS('window.flutterCanvasKitLoaded')
+external JSPromise? get windowFlutterCanvasKitLoaded;
```
Comment (ditman) [inline lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart:47]: Do we need both? `flutterCanvasKitLoaded` will eventually resolve to the same value as `windowFlutterCanvasKit`.

Can we remove the CanvasKit getter and just always await for the promise?

**Review by ditman — COMMENTED** [review]
Mostly nitpicks, I think only `createWasmInstantiator` needs a little bit more love.

File: lib/web_ui/flutter_js/src/canvaskit_loader.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    const useChromiumCanvasKit = supportsChromiumCanvasKit && (config.canvasKitVariant !== "full");
+    let baseUrl = config.canvasKitBaseUrl ?? `https://www.gstatic.com/flutter-canvaskit/${engineRevision}/`;
+    if (useChromiumCanvasKit) {
+      baseUrl = `${baseUrl}/chromium/`;
+    }
+    let canvasKitUrl = `${baseUrl}canvaskit.js`;
+    if (deps.flutterTT.policy) {
+      canvasKitUrl = deps.flutterTT.policy.createScriptURL(canvasKitUrl);
+    }
+    const wasmInstantiator = createWasmInstantiator(`${baseUrl}canvaskit.wasm`);
```
[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:23]: (Unless `''` is a valid `baseUrl`, I guess)

File: lib/web_ui/flutter_js/src/canvaskit_loader.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    const script = document.createElement("script");
+    script.src = canvasKitUrl;
+    script.addEventListener('load', async () => {
+      try {
+        const canvasKit = await CanvasKitInit({
+          instantiateWasm: wasmInstantiator,
+        });
+        window.flutterCanvasKit = canvasKit;
```
[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:31]: If we always end up overwriting flutterCanvasKit, should `flutterCanvasKit` be always a Promise and we teach the engine that it needs to always be awaited?

File: lib/web_ui/flutter_js/src/canvaskit_loader.js [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import { engineRevision } from "./revision.js";
+import { createWasmInstantiator } from "./instantiate_wasm.js";
+
+export const loadCanvasKit = (deps, config, browserEnvironment) => {
+  window.flutterCanvasKitLoaded = new Promise((resolve, reject) => {
+    const supportsChromiumCanvasKit = browserEnvironment.hasChromiumBreakIterators && browserEnvironment.hasImageCodecs;
+    if (!supportsChromiumCanvasKit && config.canvasKitVariant == "chromium") {
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:11]: Dangit JavaScript...

File: lib/web_ui/flutter_js/src/canvaskit_loader.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+      throw "Chromium CanvasKit variant specifically requested, but unsupported in this browser";
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:12]: This is intentional. If we do `reject`, then this failure won't happen until way down the road when the dart code tries to await the promise. However, we know immediately and synchronously that loading will fail, and I'd rather the bootstrapper just bail out immediately.

File: lib/web_ui/flutter_js/src/canvaskit_loader.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:23]: I originally had it the way you're suggesting, because largely I agree with you, but that's not how our code actually works right now. The `canvasKitBaseUrl` is expected to have a trailing slash currently. We need to respect the current semantics.

File: lib/web_ui/flutter_js/src/canvaskit_loader.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    let baseUrl = config.canvasKitBaseUrl ?? `https://www.gstatic.com/flutter-canvaskit/${engineRevision}/`;
+    if (useChromiumCanvasKit) {
+      baseUrl = `${baseUrl}/chromium/`;
+    }
+    let canvasKitUrl = `${baseUrl}canvaskit.js`;
+    if (deps.flutterTT.policy) {
+      canvasKitUrl = deps.flutterTT.policy.createScriptURL(canvasKitUrl);
+    }
+    const wasmInstantiator = createWasmInstantiator(`${baseUrl}canvaskit.wasm`);
+    const script = document.createElement("script");
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:24]: Yep, you're right, I'll change this.

File: lib/web_ui/flutter_js/src/canvaskit_loader.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+      } catch (e) {
+        reject(e);
+      }
+    });
+    script.addEventListener('error', (e) => {
+      reject(e);
+    });
+    document.head.appendChild(script);
+  });
+  return window.flutterCanvasKitLoaded;
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:42]: I'm not sure what you mean. It should never be null, it's a promise that is constructed immediately above. It is true that the promise isn't actually used by JavaScript (at least right now) and it's only used by the dart code later on initialization, but I figured returning the promise is probably a bit more flexible. I could be convinced otherwise though if you think returning it is a bad idea.

File: lib/web_ui/flutter_js/src/entrypoint_loader.js [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+   * @param {Function} engineInitializer @see https://github.com/flutter/engine/blob/main/lib/web_ui/lib/src/engine/js_interop/js_loader.dart#L42
+   */
+  didCreateEngineInitializer(engineInitializer) {
+    if (typeof this._didCreateEngineInitializerResolve === "function") {
+      this._didCreateEngineInitializerResolve(engineInitializer);
+      // Remove the resolver after the first time, so Flutter Web can hot restart.
+      this._didCreateEngineInitializerResolve = null;
+      // Make the engine revert to "auto" initialization on hot restart.
+      delete _flutter.loader.didCreateEngineInitializer;
+    }
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/entrypoint_loader.js:92]: My take on this is that I don't actually want to change the behavior of `loadEntrypoint` at all with this change, if possible. The reason is that I think the next step for this API should be removal after deprecation. I don't really want to change the behavior of the API and then remove it. I think it would be better to make users migrate a single time and make a clean break, rather than subtly change the behavior of the deprecated API. If users are still using the promise-based method on `loadEntrypoint`, they will be gently nudged to move over to the `load` API (where they won't rely on promise-based behavior) and then maybe one more release later we'll forcibly remove it. At that point we can get rid of this weird code path. But for now I don't think we should mess with the existing behavior of the deprecated API.

File: lib/web_ui/flutter_js/src/instantiate_wasm.js (hunk 4 above)
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/instantiate_wasm.js:9]: It certainly would be nice if this worked the way you're describing, but this is meant to be plugged into an emscripten API (`instantiateWasm`) which doesn't work with promises. I wish they did, I'm not really sure why their API isn't promise-based, but the (admittedly very strange) contract of the API is that the function takes a `successCallback`, and you can either synchronously return the module instance, or return an empty JS object (`{}`) and then call `successCallback` later asynchronously with the instance and module. Weird, I know, but it's outside of our control.

See https://emscripten.org/docs/api_reference/module.html#Module.instantiateWasm

File: lib/web_ui/flutter_js/src/loader.js (hunk 5 above)
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/loader.js:28]: I contemplated (and made an attempt at) doing this, but it's complicated because they are definitely subtly different in ways that are hard to reconcile cleanly. For example, `loadEntrypoint` only works if it knows about the build configuration, but `loadEntrypoint` actually doesn't know anything about certain parts of the build configuration (i.e. what renderer this build is using). So `loadEntrypoint` would have to make up a bogus build configuration (like saying it's the `html` renderer in order to prevent `load` from attempting to pre-initialize CanvasKit and so on) and it just feels weird and wrong. In general, I thought maybe just to keep things stable, I would try to keep this code path mostly only mechanically refactored, rather than rewritten. Hopefully we can remove it before too long anyway.

File: lib/web_ui/flutter_js/src/skwasm_loader.js (hunk 6 above)
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/skwasm_loader.js:32]: Thank god for that same-origin script policy for workers, right? :P

File: lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart (hunk 8 above)
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/canvaskit/canvaskit_api.dart:47]: I think as long as `loadEntrypoint` exists, we'll need both. The `windowFlutterCanvasKitLoaded` promise only exists when `load` is called. In the `loadEntrypoint` case, on hot restart we just want to reuse the `windowFlutterCanvasKit` that was already set by the loading code (that happened in Dart).

Once `loadEntrypoint` is removed, a bunch of this code can be simplified quite a bit (in fact, I think we can remove all the dart-side CanvasKit loading code completely at that point).

File: lib/web_ui/flutter_js/src/canvaskit_loader.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:31]: Yeah actually I think you're right. If the user sets `window.flutterCanvasKit` before they call the bootstrapper, we probably should still respect that, for now at least. I thought the main reason for this global variable was to not have to do the full canvaskit download and re-initialization on hot reload, but I could also see this being used by people to have some other side-loaded way of providing CanvasKit.

File: lib/web_ui/flutter_js/src/canvaskit_loader.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:15]: OK, since this is not inside a callback or something like that, it really doesn't matter (this will also be thrown in the await?)

File: lib/web_ui/flutter_js/src/canvaskit_loader.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:46]: Let me rephrase, because I can't even understand what I'm asking above:

Is it OK that if `loadCanvasKit` is called multiple times we create (and overwrite) the `window.flutterCanvasKitLoaded` promise multiple times, or should we always return the promise that was created the first time the method was called?

File: lib/web_ui/flutter_js/src/entrypoint_loader.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (ditman) [reply lib/web_ui/flutter_js/src/entrypoint_loader.js:92]: OK, let's keep for a while longer then!

File: lib/web_ui/flutter_js/src/loader.js (hunk 5 above)
Comment (ditman) [reply lib/web_ui/flutter_js/src/loader.js:28]: ACK, thanks for considering it though!

File: lib/web_ui/flutter_js/src/skwasm_loader.js (hunk 6 above)
Comment (ditman) [reply lib/web_ui/flutter_js/src/skwasm_loader.js:34]: The Security Sandbox (hopefully they don't close this :P)

File: lib/web_ui/flutter_js/src/instantiate_wasm.js (hunk 4 above)
Comment (ditman) [reply lib/web_ui/flutter_js/src/instantiate_wasm.js:9]: Awwwgh okay!

File: lib/web_ui/flutter_js/src/canvaskit_loader.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:15]: Yes, it will be thrown immediately and synchronously whether this is called synchronously or asynchronously.

File: lib/web_ui/flutter_js/src/browser_environment.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+const isBlink = () => {
+  return (navigator.vendor === 'Google Inc.') ||
```
Comment (ditman) [inline lib/web_ui/flutter_js/src/browser_environment.js:6]: This is very nitpicky, and can be done later, but we should pick a quote style and stick to it; either `'` or `"` (line 25) ((back-quotes are fine, I kinda prefer `"` for JS code))

File: lib/web_ui/flutter_js/src/browser_environment.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+const hasChromiumBreakIterators = () => {
+  return (typeof Intl.v8BreakIterator !== "undefined") &&
+    (typeof Intl.Segmenter !== "undefined");
+}
+
+const supportsWasmGC = () => {
+  // This attempts to instantiate a wasm module that only will validate if the
+  // final WasmGC spec is implemented in the browser.
+  //
+  // Copied from https://github.com/GoogleChromeLabs/wasm-feature-detect/blob/main/src/detectors/gc/index.js
```
Comment (ditman) [inline lib/web_ui/flutter_js/src/browser_environment.js:33]: Hmmm 🤔

File: lib/web_ui/flutter_js/src/canvaskit_loader.js [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+    if (!supportsChromiumCanvasKit && config.canvasKitVariant == "chromium") {
+      throw "Chromium CanvasKit variant specifically requested, but unsupported in this browser";
+    }
+    const useChromiumCanvasKit = supportsChromiumCanvasKit && (config.canvasKitVariant !== "full");
+    let baseUrl = config.canvasKitBaseUrl ?? `https://www.gstatic.com/flutter-canvaskit/${engineRevision}/`;
+    if (useChromiumCanvasKit) {
+      baseUrl = `${baseUrl}/chromium/`;
+    }
+    let canvasKitUrl = `${baseUrl}canvaskit.js`;
+    if (deps.flutterTT.policy) {
```
Comment (ditman) [inline lib/web_ui/flutter_js/src/canvaskit_loader.js:23]: QQ: how do we decide what goes in `deps` vs what is passed as a parameter to the function?

For example, the `browserEnvironment` looks like something that could be a `dep` too, and not need its own specific positional parameter in the `loadCanvasKit` function?

File: lib/web_ui/flutter_js/src/canvaskit_loader.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:46]: Ah, I see. Good question. I think it would be weird (by which I mean wrong) for the user to call load multiple times. Maybe it would actually make sense to explicitly throw if they did that? Or, as you say, we could just return the promise and do nothing.

File: lib/web_ui/flutter_js/src/types.d.ts [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+  mainJsPath: string;
+}
+
+export interface WasmApplicationBuild extends ApplicationBuildBase {
+  compileTarget: WasmCompileTarget;
+  mainWasmPath: string;
+  jsSupportRuntimePath: string;
+}
+
+export type ApplicationBuild = JSApplicationBuild | WasmApplicationBuild;
```
Comment (ditman) [inline lib/web_ui/flutter_js/src/types.d.ts:30]: Do you need this union? Wouldn't it be equivalent that `interface ApplicationBuildBase` is called `interface ApplicationBuild`?

File: lib/web_ui/test/fallbacks/fallbacks_test.dart [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+
+Future<void> testMain() async {
+  setUpUnitTests(
+    setUpTestViewDimensions: false,
+  );
+
+  test('bootstrapper selects correct builds', () {
+    if (browserEngine == BrowserEngine.blink) {
+      expect(isWasm, isTrue);
+      expect(isSkwasm, isTrue);
```
Comment (ditman) [inline lib/web_ui/test/fallbacks/fallbacks_test.dart:25]: Is there any case where `isWasm == true` but `isSkwasm == false`?

**Review by ditman — APPROVED** [review]
Thanks for this, this is a fantastic improvement! (All my comments are minor/questions, feel free to disregard!)

File: lib/web_ui/flutter_js/src/types.d.ts (hunk 19 above)
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/types.d.ts:30]: A union like this is super useful in TS. An `interface` basically has open polymorphism, but a union has closed polymorphism. This basically means that if you're using the union (`ApplicationBuild`), you can actually differentiate based on the `compileTarget` field, and the compiler will actually refine the type in the the subsequent branches. This won't happen when using the interface (and in fact, the `compileTarget` field doesn't exist on the interface, so you couldn't check for it). Example:
```
const build: ApplicationBuild = //something;
// build is an ApplicationBuild type
if (build.compileTarget === "dart2wasm") {
  // build is actually a WasmApplicationBuild in this branch, so I can access fields specific to WasmApplicationBuild
  console.log(build.mainWasmPath);
} else {
  // the compiler can even tell that the only remaining possible values are of type JSApplicationBuild
  console.log(build.mainJsPath);
}
```

File: lib/web_ui/test/fallbacks/fallbacks_test.dart (hunk 20 above)
Comment (eyebrowsoffire) [reply lib/web_ui/test/fallbacks/fallbacks_test.dart:25]: Not in this test suite. In this test suite, we just have two build variants, skwasm + dart2wasm and dart2js + canvaskit.

File: lib/web_ui/flutter_js/src/browser_environment.js (hunk 16 above)
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/browser_environment.js:6]: Yes sorry I tried to be consistent with double quotes in JS code but missed some spots.

File: lib/web_ui/flutter_js/src/canvaskit_loader.js (hunk 18 above)
Comment (eyebrowsoffire) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:23]: I guess it's mostly arbitrary. The browser environment just logically felt like something different than a "dependency" so I have it as a separate thing. Perhaps I don't even need the `browserEnvironment` to be passed as a parameter anymore now that it's just a global variable vended by the `browser_environment.js` module...

Comment (ditman) [reply lib/web_ui/flutter_js/src/canvaskit_loader.js:23]: > now that it's just a global variable vended by the browser_environment.js module

I thought the same, but also if you grab it from deps (or the parameter) it's easy(er) to mock for testing maybe?

**Comment by bdero** [comment]
FYI @eyebrowsoffire, this patch regressed a couple of size benchmarks: https://flutter-flutter-perf.skia.org/e/?keys=Xe5ab9093da633e5f5a62ea9d8f4fa9f6&selected=commit%3D38703%26name%3D%252Carch%253Dintel%252Cbranch%253Dmaster%252Cconfig%253Ddefault%252Cdevice_type%253Dnone%252Cdevice_version%253Dnone%252Chost_type%253Dlinux%252Csub_result%253Dhello_world_flutter_js_uncompressed_bytes%252Ctest%253Dweb_size__compile_test%252C&xbaroffset=38703

In particular, hello_world_flutter_js_uncompressed_bytes doubled. I'm not intimately familiar with these, so not sure if this is expected behavior or not.

**Comment by eyebrowsoffire** [comment]
@bdero Definitely expected. Thanks for the heads up though!

**Comment by rafaelbenedetti1** [comment]
I'm having a problem after update my flutter version to 3.22, I modify the index.html to new mode of app initialization. My application have some textfields, and after upgrade the number keyboard doesn't work, because the six and four keys moves the cursor to right and left and not show the number on field, how can i resolve this?

**Comment by eyebrowsoffire** [comment]
@rafaelbenedetti1 Could you file a separate issue with a minimum reproducible case?

**Comment by rafaelbenedetti1** [comment]
@eyebrowsoffire  where can i create?

**Comment by eyebrowsoffire** [comment]
@rafaelbenedetti1  https://github.com/flutter/flutter/issues/new/choose

**Comment by rafaelbenedetti1** [comment]
@eyebrowsoffire https://github.com/flutter/flutter/issues/150191 here
